docs/proposals/container-init.md

```diff
@@ -0,0 +1,333 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Pod initialization
+
+@smarterclayton
+
+March 2016
+
+## Proposal and Motivation
+
+Within a pod there is a need to initialize local data or adapt to the current
+cluster environment that is not easily achieved in the current container model.
+Containers start in parallel after volumes are mounted, leaving no opportunity
+for coordination between containers without specialization of the image. If
+two containers need to share common initialization data, both images must
+be altered to cooperate using filesystem or network semantics, which introduces
+coupling between images. Likewise, if an image requires configuration in order
+to start and that configuration is environment dependent, the image must be
+altered to add the necessary templating or retrieval.
+
+This proposal introduces the concept of an **init container**, one or more
+containers started in sequence before the pod's normal containers are started.
+These init containers may share volumes, perform network operations, and perform
+computation prior to the start of the remaining containers. They may also, by
+virtue of their sequencing, block or delay the startup of application containers
+until some precondition is met. In this document we refer to the existing pod
+containers as **app containers**.
+
+## Design Points
+
+* Init containers should be able to:
+  * Perform initialization of shared volumes
```

Comment:
For shared volumes do are you concerned that this proposal would
block, limit, or complicate the efficacy / use of shared volumes?  I
would view the container volume as additive on top of this, but with
some orthogonal concerns this proposal does not require changes to
address (long running "init" container, implicit start order,
precondition order, may require chain restarts on docker if the
container provider is restarted, much more complex changes to the
kubelet and volume code).

Container volumes (vs image volumes) require a virtual DAG
implementation for restart, except that users cannot set
preconditions.  Would container volumes that run forever (fuse
providing volume) be considered an init container?  I'd argue it does
not sound like that, so if it was a regular container we would have a
limited DAG in the pod containers spec.  If we have a limited DAG,
users can then construct deeper dependencies and abuse the DAG.